BDArmory/Weapons/Missiles/MissileBase.cs

```diff
@@ -1629,11 +1636,54 @@ public void CheckDetonationState(bool separateWarheads = false)
                                             catch (Exception e)
                                             {
                                                 // ignored
-                                                Debug.LogWarning("[BDArmory.MissileBase]: Exception thrown in CheckDetonatationState: " + e.Message + "\n" + e.StackTrace);
+                                                Debug.LogWarning("[BDArmory.MissileBase]: Exception thrown in CheckDetonationState: " + e.Message + "\n" + e.StackTrace);
                                             }
                                         }
                                     }
                                 }
+                                else if (TargetAcquired && targetVessel != null && targetVessel.Vessel != null)
+                                {
+                                    // For very high speed intercepts when missiles may phase through small vessels/missiles within a frame
+                                    Vector3 relPos = TargetPosition - vessel.CoM;
+                                    Vector3 relVel = TargetVelocity - vessel.Velocity();
+                                    bool approaching = Vector3.Dot(relPos, relVel) < 0f;
+                                    float targetRad = targetVessel.Vessel.GetRadius();
+                                    float selfRad = vessel.GetRadius();
+                                    float sepRad = (targetRad + selfRad);
+
+                                    if (approaching && relVel.sqrMagnitude * Time.fixedDeltaTime * Time.fixedDeltaTime > sepRad * sepRad)
+                                    {
+                                        try
+                                        {
+                                            Vector3 relAccel = TargetAcceleration - vessel.acceleration_immediate;
+                                            float cpaTime = AIUtils.TimeToCPA(relPos, relVel, relAccel, Time.fixedDeltaTime);
+
+                                            if (cpaTime < Time.fixedDeltaTime && cpaTime > 0f)
+                                            {
+                                                float sqrSep = AIUtils.PredictPosition(relPos, relVel, relAccel, cpaTime).sqrMagnitude;
+                                                if (sepRad * sepRad > sqrSep)
+                                                {
+                                                    // Valid hit, determine hit point and set transform.position (moves immediately) to hit point
+                                                    Vector3 missileToTarget = AIUtils.PredictPosition(TargetPosition, TargetVelocity, TargetAcceleration, cpaTime) - vessel.PredictPosition(cpaTime);
+                                                    if (Physics.Raycast(new Ray(TargetPosition - sepRad * missileToTarget, missileToTarget), out RaycastHit hit, sepRad * sepRad, (int)(LayerMasks.Parts | LayerMasks.EVA | LayerMasks.Wheels)))
```

Comment:
`sepRad` is the sum of the largest dimensions of the two craft, so starting `sepRad` distance away from `TargetPosition` should put it beyond colliders almost all of the time. Also, it shouldn't be close to zero unless the missile and target are miniscule. Are you mixing it up with `sqrSep`, the CPA distance squared, perhaps?
You could use `1.7321f * sepRad` (i.e., marginally over `sqrt(3)` times the sum of the largest dimensions) to be sure since that'd be the largest an object could possibly be going between diagonally opposite corners of the cube with edge size `sepRad`.

In any case, the raycast needs to be the same length as whatever distance away from `TargetPosition` that you're starting it otherwise it won't reach `TargetPosition`, so starting at `TargetPosition - sepRad * missileToTarget` would require a cast length of `sepRad * missileToTarget.magnitude`, not `sepRad * sepRad`. I don't think you can avoid normalising a vector here as you need to know both the distance and the direction. However, you can compute it only once by taking advantage of the fact that `Ray` is going to normalise the direction in it's constructor anyway by doing
```csharp
float targetRad = 1.7321f * targetVessel.Vessel.GetRadius(); // Maximum possible radius of the target.
float selfRad = 1.7321f * vessel.GetRadius();
float sepRad = targetRad + selfRad;
...
Ray ray = new(TargetPosition, missileToTarget);
ray.origin -= sepRad * ray.direction;
if (Physics.Raycast(ray, out RaycastHit hit, sepRad, (int)(LayerMasks.Parts | LayerMasks.EVA | LayerMasks.Wheels)))
```

One other thing to note is that if the missile is actually very close to the target when you do the raycast (i.e., `cpaTime` is almost 0), then it's possible the ray could be originating on the other side of the missile and the raycast will then hit the missile instead of the target. This probably wouldn't make much difference though as the missile would be close enough that detonating in its current position would still damage the target roughly the same amount. However, using `targetRad` instead of `sepRad` in the ray and raycast would reduce the chances of this happening since the ray only needs to start beyond the bounds of the target to be able to hit its colliders.

Reply:
Thinking about this some more, I think the raycast is doing the wrong thing in any case. It shouldn't be raycasting from the direction of the CPA point as that's not the direction that the missile is approaching from. Instead something like the following should work (I think, I haven't tested it):
```diff
diff --git a/BDArmory/Extensions/VesselExtensions.cs b/BDArmory/Extensions/VesselExtensions.cs
index d2c39ae61..9f1a12a12 100644
--- a/BDArmory/Extensions/VesselExtensions.cs
+++ b/BDArmory/Extensions/VesselExtensions.cs
@@ -109,7 +109,7 @@ public static float GetRadius(this Vessel vessel, Vector3 fireTransform = defaul
 #if DEBUG
                 if (radius < bounds.x / 2f && radius < bounds.y / 2f && radius < bounds.z / 2f) Debug.LogWarning($"DEBUG Radius {radius} of {vessel.vesselName} is less than half its minimum bounds {bounds}");
 #endif
-                return Mathf.Min(radius, (Mathf.Max(Mathf.Max(vessel.vesselSize.x, vessel.vesselSize.y), vessel.vesselSize.z) / 2f) * 1.732f); // clamp bounds to vesselsize in case of Bounds erroneously reporting vessel sizes that are impossibly large
+                return Mathf.Min(radius, (Mathf.Max(Mathf.Max(vessel.vesselSize.x, vessel.vesselSize.y), vessel.vesselSize.z) / 2f) * 1.7321f); // clamp bounds to vesselsize in case of Bounds erroneously reporting vessel sizes that are impossibly large
             }
         }
 
diff --git a/BDArmory/Weapons/Missiles/MissileBase.cs b/BDArmory/Weapons/Missiles/MissileBase.cs
index f7bf42836..4f164f835 100644
--- a/BDArmory/Weapons/Missiles/MissileBase.cs
+++ b/BDArmory/Weapons/Missiles/MissileBase.cs
@@ -1520,7 +1520,7 @@ public void CheckDetonationState(bool separateWarheads = false)
             var futureTargetPosition = (TargetPosition + targetDistancePerFrame);
             var futureMissilePosition = (vessel.CoM + missileDistancePerFrame);
 
-            var relativeSpeed = (TargetVelocity - vessel.Velocity()).magnitude * Time.fixedDeltaTime;
+            float relativeSpeed = (float)(TargetVelocity - vessel.Velocity()).magnitude * Time.fixedDeltaTime; // relativeSpeed is actually a distance!
 
             switch (DetonationDistanceState)
             {
@@ -1649,27 +1649,32 @@ public void CheckDetonationState(bool separateWarheads = false)
                                     bool approaching = Vector3.Dot(relPos, relVel) < 0f;
                                     float targetRad = targetVessel.Vessel.GetRadius();
                                     float selfRad = vessel.GetRadius();
-                                    float sepRad = (targetRad + selfRad);
+                                    float sepRad = 1.7321f * (targetRad + selfRad);
 
                                     if (approaching && relVel.sqrMagnitude * Time.fixedDeltaTime * Time.fixedDeltaTime > sepRad * sepRad)
                                     {
-                                        try
+                                        bool shouldDetonate = false;
+                                        Ray ray = new(vessel.CoM, relVel);
+                                        ray.origin += selfRad * ray.direction; // Start at the tip of the missile (assuming it's pointing roughly prograde in the relVel direction and is longest on that axis).
+                                        if (Physics.Raycast(ray, out RaycastHit hit, relativeSpeed, (int)(LayerMasks.Parts | LayerMasks.EVA | LayerMasks.Wheels))) // Hit!
+                                        {
+                                            vessel.SetPosition(hit.point);
+                                            shouldDetonate = true;
+                                        }
+                                        else // Not hitting, just getting close — check for reaching CPA.
                                         {
                                             Vector3 relAccel = TargetAcceleration - vessel.acceleration_immediate;
                                             float cpaTime = AIUtils.TimeToCPA(relPos, relVel, relAccel, Time.fixedDeltaTime);
-
-                                            if (cpaTime < Time.fixedDeltaTime && cpaTime > 0f)
+                                            if (cpaTime > 0f && cpaTime < Time.fixedDeltaTime)
                                             {
-                                                float sqrSep = AIUtils.PredictPosition(relPos, relVel, relAccel, cpaTime).sqrMagnitude;
-                                                if (sepRad * sepRad > sqrSep)
+                                                // Set relative position to the same as at CPA point, but relative to the target's current position. This avoids having to move the target and wait an additional frame.
+                                                vessel.SetPosition(TargetPosition - AIUtils.PredictPosition(relPos, relVel, relAccel, cpaTime));
+                                                shouldDetonate = true;
+                                            }
+                                        }
+                                        if (shouldDetonate)
                                         {
-                                                    // Valid hit, determine hit point and set transform.position (moves immediately) to hit point
-                                                    Vector3 missileToTarget = AIUtils.PredictPosition(TargetPosition, TargetVelocity, TargetAcceleration, cpaTime) - vessel.PredictPosition(cpaTime);
-                                                    if (Physics.Raycast(new Ray(TargetPosition - sepRad * missileToTarget, missileToTarget), out RaycastHit hit, sepRad * sepRad, (int)(LayerMasks.Parts | LayerMasks.EVA | LayerMasks.Wheels)))
-                                                        transform.position = hit.point;
-                                                    else // Otherwise set relative position the same as at CPA point, but relative to the target's current position. This avoids having to move the target and wait an additional frame.
-                                                        transform.position = TargetPosition - missileToTarget;
-                                                    vessel.SetPosition(transform.position);
+                                            // vessel.CoM = vessel.transform.position; // Note: SetPosition takes effect immediately, but doesn't update CoM or other fields. If CoM is needed (it doesn't appear to be), then uncomment this.
                                             Detonate();
                                             if (targetVessel.isMissile && targetVessel.MissileBaseModule)
                                                 targetVessel.MissileBaseModule.Detonate(); // The above approach is only about ~50% effective against missiles, so just tell the other missile to detonate just in case
@@ -1677,13 +1682,6 @@ public void CheckDetonationState(bool separateWarheads = false)
                                         }
                                     }
                                 }
-                                        catch (Exception e)
-                                        {
-                                            // ignored
-                                            Debug.LogWarning("[BDArmory.MissileBase]: Exception thrown in CheckDetonationState: " + e.Message + "\n" + e.StackTrace);
-                                        }
-                                    }
-                                }
                             }
                         }
                         else
```